### PR TITLE
Extension of spatial.distance module

### PR DESCRIPTION
**Pull request by Cdebus**
- refactorization to scipy naming
- restructuring of distance calculation logic
- support for distance calculation of two different tensors with different splits
- support for dtypes other than float32

## Description

Issue/s resolved: 
Addresses #475 amongst other changes

## Changes proposed:
- Implementation of abstract function _dist(X,Y, metric)
- derive cdist and rbf as explicit implementatiions with metrices euclidian and gaussian

## Type of change
- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relavent functions
- [ ] Documentation updated (if needed)
- [ ] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/spatial/distance.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+import numpy as np
+from mpi4py import MPI
+
+from ..core import factories
+from ..core import types
+
+__all__ = ["cdist", "rbf"]
+
+
+def _euclidian(x, y):
```
Comment (coquelin77) [inline heat/spatial/distance.py]: need some short docs for the internal functions and proper docs for the functions which are visible from outside (cdist and rbf)

**Comment by coquelin77** [comment]
would it be possible to abstract the communications loop that you use in for the circular sending/receiving?

File: heat/spatial/distance.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+                    count = int(stat.Get_count(mpi_type) / f)
+                    moving = torch.zeros((count, f), dtype=torch_type)
+                    comm.Recv(moving, source=sender, tag=iter)
+
+                d_ij = metric(stationary, moving)
+                d._DNDarray__array[:, columns[0] : columns[1]] = d_ij
+
+                # Receive calculated tile
+                scol1 = displ[receiver]
+                if receiver != size - 1:
```
[the remark this answers is not in the document]
Comment (Cdebus) [reply heat/spatial/distance.py]: Thanks

**Comment by Cdebus** [comment]
> 
> 
> would it be possible to abstract the communications loop that you use in `_dist` for the circular sending/receiving?

I don't know, in principle yes, but what would be the application. The key point here ist the calculation of metrics between chunks from X and Y. While this is easily substituted, it can already currently be modified by the function you pass as metric.
